# **RANDOM FOREST**

In [5]:
#LIBRERIAS
import pandas as pd
import matplotlib.pyplot as plt
from pickle import dump

from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

from sklearn.model_selection import GridSearchCV

In [2]:
#DATA
sel_train= pd.read_csv ("/workspaces/Jese_David-Random_Forest-tutorial/Jese_David-Decision-Tree-project/data/Datasets/sel_dataset_train.csv")
sel_test = pd.read_csv ("/workspaces/Jese_David-Random_Forest-tutorial/Jese_David-Decision-Tree-project/data/Datasets/sel_dataset_test.csv")
target = "Outcome"

#Split
X_sel_train = sel_train.drop(columns=[target])
y_sel_train = sel_train[target]

X_sel_test = sel_test.drop(columns=[target])
y_sel_test = sel_test[target]

### **RANDOM FOREST BASE**

In [3]:
rf_sel = RandomForestClassifier(
    n_estimators=200,        # nº de árboles
    max_depth=None,          # sin limitar de momento
    random_state=42,
    class_weight="balanced", # por el desbalance de Outcome
    n_jobs=-1                # usa todos los cores
)

rf_sel.fit(X_sel_train, y_sel_train)

y_sel_pred_rf = rf_sel.predict(X_sel_test)

print("\n=== RANDOM FOREST (SEL) ===")
print("Accuracy test:", accuracy_score(y_sel_test, y_sel_pred_rf))
print("\nMatriz de confusión:\n", confusion_matrix(y_sel_test, y_sel_pred_rf))
print("\nReporte de clasificación:\n", classification_report(y_sel_test, y_sel_pred_rf))



=== RANDOM FOREST (SEL) ===
Accuracy test: 0.7207792207792207

Matriz de confusión:
 [[80 20]
 [23 31]]

Reporte de clasificación:
               precision    recall  f1-score   support

           0       0.78      0.80      0.79       100
           1       0.61      0.57      0.59        54

    accuracy                           0.72       154
   macro avg       0.69      0.69      0.69       154
weighted avg       0.72      0.72      0.72       154



### **Guardo Modelo**

In [6]:
dump(
    rf_sel,
    open("../models/RandomForest_sel.sav", "wb")
)


1. **Árbol base** (SEL)

Accuracy: 0.68

Matriz: [[75 25], [24 30]]

Clase 1 (diabetes):

Precision: 0.55

Recall: 0.56

F1: 0.55

Modelo justito: ni mucha accuracy ni buen recall en la clase 1.

2. **Random Forest** (SEL)

Accuracy: 0.72

Matriz: [[80 20], [23 31]]

Clase 1:

Precision: 0.61

Recall: 0.57

F1: 0.59

Mejora un poco al árbol base:

Sube accuracy (0.68 → 0.72)

Ligera mejora en F1 de la clase 1 (0.55 → 0.59)
Pero el recall de la clase 1 sigue alrededor de 0.57 → sigues perdiendo muchos diabéticos.

3. **Árbol optimizado** (SEL)

Accuracy: 0.75

Matriz: [[69 31], [ 8 46]]

Clase 1:

Precision: 0.60

Recall: 0.85

F1: 0.70

Aquí es donde se ve el salto fuerte:

Accuracy sube a 0.75 (mejor que RF).

La clase 1 mejora mucho:

recall: 0.57 → 0.85

F1: 0.59 → 0.70

Ese recall alto con más falsos positivos (31 en vez de 20), pero en un contexto de salud eso es aceptable: prefieres revisar a más gente “sospechosa” antes que dejar escapar a diabéticos reales.